### PR TITLE
History - `!!` and simple `!`

### LINKED ISSUE #40: feat: simple history
- [x] `history`
- [x] `!` with number
- [x] `!!` for last command

### PR DESCRIPTION
**Pull request by RezaRahemtola**
# Description

History-related elements with `!!` to use the last command and simple `!` implementation to use the history.

Closes #40 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the needed labels
- [x] I have linked this PR to an issue
- [x] I have linked this PR to a milestone
- [ ] I have linked this PR to a project
- [ ] I have tested this code
- [ ] I have added / updated tests (unit / functionals / end-to-end / ...)
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: src/utils/history/history.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#include "history.h"
+#include "my_string.h"
+#include "my_math.h"
+
+static void replace_last_command(char **input, list_t *history)
+{
+    char *new = NULL;
+
+    if (my_list_size(history) == 0) {
+        dprintf(2, "0: Event not found.\n");
```
Comment (Croos3r) [inline src/utils/history/history.c]: fprintf could be more readable

File: src/utils/history/history.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    while (new != NULL && *input != NULL) {
+        if (isdigit(new[1]))
+            handle_nb_substitution(input, &new[1], history);
+        new = strchr(++new, '!');
+    }
+}
+
+void replace_history(char **input, shell_t *shell)
+{
+    char *old = strdup(*input);
```
Comment (Croos3r) [inline src/utils/history/history.c:29]: You check if *input is not null, but if it is, doesn't make it not working (even crash) ?

Comment (Croos3r) [reply src/utils/history/history.c:29]: Cuz you make strchr and strstr

Comment (Croos3r) [reply src/utils/history/history.c:29]: Even if it doesn't crash prefer handling this special case first

Comment (RezaRahemtola) [reply src/utils/history/history.c:29]: > You check if *input is not null, but if it is, doesn't make it not working (even crash) ?

Replacing it with " " at the end if it‘s NULL so it won’t crash

Comment (RezaRahemtola) [reply src/utils/history/history.c:29]: > Cuz you make strchr and strstr

NULL checks are done just before calling these functions

**Pull request merged**